In [2]:
import kagglehub
path = kagglehub.dataset_download("oleksiimartusiuk/e-commerce-data-shein")

100%|██████████| 3.44M/3.44M [00:00<00:00, 145MB/s]

Extracting files...


In [3]:
import pandas as pd
import numpy as np

In [4]:
print(path)

/root/.cache/kagglehub/datasets/oleksiimartusiuk/e-commerce-data-shein/versions/4


In [5]:
import os
# List all files and folders in the downloaded path

count = 0
files = os.listdir(path)
print("\nFiles in directory:")
for file in files:
    print(file)
    count = count+1

print(count)


Files in directory:
us-shein-tools_and_home_improvement-3903.csv
us-shein-toys_and_games-3577.csv
us-shein-home_textile-3883.csv
us-shein-jewelry_and_accessories-3548.csv
us-shein-beauty_and_health-4267.csv
us-shein-bags_and_luggage-4299.csv
us-shein-underwear_and_sleepwear-4019.csv
us-shein-electronics-4395.csv
us-shein-curve-2849.csv
us-shein-automotive-4110.csv
us-shein-sports_and_outdoors-3853.csv
us-shein-pet_supplies-4083.csv
us-shein-baby_and_maternity-4433.csv
us-shein-home_and_kitchen-3719.csv
us-shein-appliances-3987.csv
us-shein-swimwear-3761.csv
us-shein-womens_clothing-4620.csv
us-shein-office_and_school_supplies-4233.csv
us-shein-kids-4314.csv
us-shein-shoes-4381.csv
us-shein-mens_clothes-1891.csv
21


In [6]:
df_appliances = pd.read_csv(path + '/us-shein-appliances-3987.csv')
df_appliances.head(5)

,goods-title-link--jump,goods-title-link--jump href,rank-title,rank-sub,price,discount,selling_proposition,goods-title-link
0,1pc Rechargeable Deep Tissue Muscle Handheld M...,https://us.shein.com/1pc-Rechargeable-Deep-Tis...,#1 Best Sellers,in Give Gifts,$2.03,-22%,NaN,NaN
1,1pc Portable Hanging Neck Fan,https://us.shein.com/1pc-Portable-Hanging-Neck...,#4 Best Sellers,in Top rated in Portable Fans,$6.48,-20%,NaN,NaN
2,1pc Pink Colored Curved Eyelash Curler False E...,https://us.shein.com/1pc-Pink-Colored-Curved-E...,NaN,NaN,$1.80,NaN,400+ sold recently,NaN
3,1 Mini Portable Handheld Fan With 2 Aa Batteri...,https://us.shein.com/1-Mini-Portable-Handheld-...,NaN,NaN,$0.88,-72%,5.6k+ sold recently,NaN
4,"Wit Water Flosser,Portable Oral Irrigator With...",https://us.shein.com/Wit-Water-Flosser-Portabl...,#6 Best Sellers,in Oral Irrigators,$12.06,-40%,NaN,NaN


In [7]:
df_office_and_school_supplies = pd.read_csv(path + '/us-shein-office_and_school_supplies-4233.csv')
df_office_and_school_supplies.head(5)

,goods-title-link--jump,goods-title-link--jump href,rank-title,rank-sub,price,discount,selling_proposition,goods-title-link
0,"Colors Dual Tip Alcohol Based Art Markers, 262...",https://us.shein.com/Colors-Dual-Tip-Alcohol-B...,#4 Best Sellers,in Markers & Highlighters,$6.88,-20%,NaN,NaN
1,1pc Expandable Large Capacity Canvas Simple Pe...,https://us.shein.com/1pc-Expandable-Large-Capa...,NaN,NaN,$4.80,-20%,1.0k+ sold recently,NaN
2,1pc Budget Planner - Monthly Finance Organizer...,https://us.shein.com/1pc-Budget-Planner-Monthl...,#1 Best Sellers,in Planners,$2.11,-67%,NaN,NaN
3,1pc Letter Graphic Plan Board,https://us.shein.com/1pc-Letter-Graphic-Plan-B...,#3 Best Sellers,in Calendars,$0.93,-77%,NaN,NaN
4,1pc Stylish And Practical Multi-Function Trave...,https://us.shein.com/1pc-Stylish-And-Practical...,NaN,NaN,$1.90,NaN,10k+ sold recently,NaN


In [8]:
count = 0
all_dfs = []
for file in files:
    if file.endswith('.csv'):
      df = pd.read_csv(path + '/' + file)
      all_dfs.append(df)
      count = count + 1

print(count)

21


## 1. Merging Files (`pd.concat`)
* **What it does:** Combines multiple DataFrames vertically into a single master table.
* **`ignore_index=True`:** Crucial. Resets row numbers to a single continuous index (e.g., `0` to `82104`), preventing duplicated row labels.
* **Schema Alignment:** Automatically handles files with different numbers of columns. It maps identical columns together and creates empty `NaN` rows for files missing those attributes.

In [25]:
df_master = pd.concat(all_dfs, ignore_index=True)
print(df_master.shape)
print("82105: The total number of individual products across all 21 files.\n"
"12: The total number of attributes (columns) describing each product. Notice how it went from 8 or 9 columns originally to 12? That’s because some files had slight structural differences or extra columns, and pd.concat gracefully aligned all of them for you!")

(82105, 12)
82105: The total number of individual products across all 21 files.
12: The total number of attributes (columns) describing each product. Notice how it went from 8 or 9 columns originally to 12? That’s because some files had slight structural differences or extra columns, and pd.concat gracefully aligned all of them for you!


In [26]:
df_master.head(10)

,goods-title-link--jump,goods-title-link--jump href,rank-title,rank-sub,price,selling_proposition,discount,goods-title-link,color-count,blackfridaybelts-bg src,blackfridaybelts-content,product-locatelabels-img src
0,46pcs Car Repair Tool Kit: Ratchet Torque Wren...,https://us.shein.com/46pcs-Car-Repair-Tool-Kit...,#10 Best Sellers,in Hand Tools,$1.10,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,"1pc Creative Toilet Sticker, Please Rinse, Fun...",https://us.shein.com/1pc-Creative-Toilet-Stick...,NaN,NaN,$0.33,700+ sold recently,-76%,NaN,NaN,NaN,NaN,NaN
2,"1pc Wall Mounted Hat Rack, Self Adhesive Hat H...",https://us.shein.com/1pc-Wall-Mounted-Hat-Rack...,NaN,NaN,$1.70,3.7k+ sold recently,-15%,NaN,NaN,NaN,NaN,NaN
3,"1pc Solar Garden Watering Can Lights,Solar Wat...",https://us.shein.com/1pc-Solar-Garden-Watering...,#4 Best Sellers,in Path Lights,$4.38,NaN,-64%,NaN,NaN,NaN,NaN,NaN
4,1000/500/100/50/20pcs Disposable Nitrile Glove...,https://us.shein.com/1000-500-100-50-20pcs-Dis...,NaN,NaN,$3.10,4.2k+ sold recently,NaN,NaN,NaN,NaN,NaN,NaN
5,3pcs Creative Powerful Mop Hook Shelf Traceles...,https://us.shein.com/3pcs-Creative-Powerful-Mo...,NaN,NaN,$0.90,NaN,-50%,NaN,NaN,NaN,NaN,NaN
6,"16ft-100ft LED Strip Lights,Sync With Music,Wi...",https://us.shein.com/16ft-100ft-LED-Strip-Ligh...,NaN,NaN,$3.70,10k+ sold recently,NaN,NaN,NaN,NaN,NaN,NaN
7,"3ft-100ft LED Strip Lights,With IR 44Key Remo...",https://us.shein.com/3ft-100ft-LED-Strip-Light...,#9 Best Sellers,in Lighting & Lamp,$3.50,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8,"3pcs Random Color Round Data Cable Organizer, ...",https://us.shein.com/3pcs-Random-Color-Round-D...,NaN,NaN,$1.28,1.8k+ sold recently,-20%,NaN,NaN,NaN,NaN,NaN
9,1pc Transparent Stainless Steel Household Scis...,https://us.shein.com/1pc-Transparent-Stainless...,#9 Best Sellers,in Hand Scissors,$0.51,NaN,-76%,NaN,NaN,NaN,NaN,NaN


## 📌 Pandas Reference: `pd.to_numeric()`

### 1. What it does
* Converts an array, Series, or column of data into a proper numeric data type (either integers or floats).
* Essential for data pipelines because regular string functions (like `.str.extract()`) output text strings, not numbers you can do math with.

### 2. The Power of `errors='coerce'`
This parameter controls what happens when Pandas encounters a value it cannot turn into a number (e.g., text like `"unknown"`, `"N/A"`, or corrupt strings):
* **`coerce`:** Safely forces all non-convertible text into a true **`NaN`** (Not a Number) blank space instead of crashing your script.
* This allows you to chain a `.fillna(0)` right afterward to cleanly replace those errors with zeros.

In [27]:
target_cols = ['rank-title', 'discount']

for col in target_cols:
    # 1. Extract digits and convert to numeric safely
    df_master[col] = pd.to_numeric(
        df_master[col].astype(str).str.extract(r'(\d+)', expand=False),
        errors='coerce'
    )
    # 2. Fill NaN with 0 and convert to integer
    df_master[col] = df_master[col].fillna(0).astype(int)

df_master.head(10)

,goods-title-link--jump,goods-title-link--jump href,rank-title,rank-sub,price,selling_proposition,discount,goods-title-link,color-count,blackfridaybelts-bg src,blackfridaybelts-content,product-locatelabels-img src
0,46pcs Car Repair Tool Kit: Ratchet Torque Wren...,https://us.shein.com/46pcs-Car-Repair-Tool-Kit...,10,in Hand Tools,$1.10,NaN,0,NaN,NaN,NaN,NaN,NaN
1,"1pc Creative Toilet Sticker, Please Rinse, Fun...",https://us.shein.com/1pc-Creative-Toilet-Stick...,0,NaN,$0.33,700+ sold recently,76,NaN,NaN,NaN,NaN,NaN
2,"1pc Wall Mounted Hat Rack, Self Adhesive Hat H...",https://us.shein.com/1pc-Wall-Mounted-Hat-Rack...,0,NaN,$1.70,3.7k+ sold recently,15,NaN,NaN,NaN,NaN,NaN
3,"1pc Solar Garden Watering Can Lights,Solar Wat...",https://us.shein.com/1pc-Solar-Garden-Watering...,4,in Path Lights,$4.38,NaN,64,NaN,NaN,NaN,NaN,NaN
4,1000/500/100/50/20pcs Disposable Nitrile Glove...,https://us.shein.com/1000-500-100-50-20pcs-Dis...,0,NaN,$3.10,4.2k+ sold recently,0,NaN,NaN,NaN,NaN,NaN
5,3pcs Creative Powerful Mop Hook Shelf Traceles...,https://us.shein.com/3pcs-Creative-Powerful-Mo...,0,NaN,$0.90,NaN,50,NaN,NaN,NaN,NaN,NaN
6,"16ft-100ft LED Strip Lights,Sync With Music,Wi...",https://us.shein.com/16ft-100ft-LED-Strip-Ligh...,0,NaN,$3.70,10k+ sold recently,0,NaN,NaN,NaN,NaN,NaN
7,"3ft-100ft LED Strip Lights,With IR 44Key Remo...",https://us.shein.com/3ft-100ft-LED-Strip-Light...,9,in Lighting & Lamp,$3.50,NaN,0,NaN,NaN,NaN,NaN,NaN
8,"3pcs Random Color Round Data Cable Organizer, ...",https://us.shein.com/3pcs-Random-Color-Round-D...,0,NaN,$1.28,1.8k+ sold recently,20,NaN,NaN,NaN,NaN,NaN
9,1pc Transparent Stainless Steel Household Scis...,https://us.shein.com/1pc-Transparent-Stainless...,9,in Hand Scissors,$0.51,NaN,76,NaN,NaN,NaN,NaN,NaN


In [28]:
df_master.head(1)

,goods-title-link--jump,goods-title-link--jump href,rank-title,rank-sub,price,selling_proposition,discount,goods-title-link,color-count,blackfridaybelts-bg src,blackfridaybelts-content,product-locatelabels-img src
0,46pcs Car Repair Tool Kit: Ratchet Torque Wren...,https://us.shein.com/46pcs-Car-Repair-Tool-Kit...,10,in Hand Tools,$1.10,NaN,0,NaN,NaN,NaN,NaN,NaN


## 📌 Code Summary: Missing Value Treatment & Type Casting

### 1. Specific Numeric Handling
* Checks if `color-count` exists and fills its blank slots (`NaN`) with `0`. This keeps it a number so you can do math with it later.

### 2. Blanket Fill (`.fillna('unknown')`)
* Fills **every remaining empty spot** in the entire dataset with the word `'unknown'` in one single shot. This ensures there are no missing data gaps left.

### 3. Text Column Enforcement
* Loops through every column in the dataset.
* If the column isn't one of your numeric ones (`color-count`, `rank-title`, `discount`), it forces it into a clean string data type (`.astype(str)`). This standardizes all text fields (like titles and URLs).

In [29]:
# 1. Fill the numeric column with 0
if 'color-count' in df_master.columns:
    df_master['color-count'] = df_master['color-count'].fillna(0)

# 2. Blanket fill EVERY other NaN in the entire dataframe with 'unknown' in one shot
df_master = df_master.fillna('unknown')

# 3. Force all text columns to be actual string data types
for col in df_master.columns:
    if col != 'color-count' and col != 'rank-title' and col != 'discount':
        df_master[col] = df_master[col].astype(str)

# 4. Print to verify the clean output
df_master.head(10)

,goods-title-link--jump,goods-title-link--jump href,rank-title,rank-sub,price,selling_proposition,discount,goods-title-link,color-count,blackfridaybelts-bg src,blackfridaybelts-content,product-locatelabels-img src
0,46pcs Car Repair Tool Kit: Ratchet Torque Wren...,https://us.shein.com/46pcs-Car-Repair-Tool-Kit...,10,in Hand Tools,$1.10,unknown,0,unknown,0.0,unknown,unknown,unknown
1,"1pc Creative Toilet Sticker, Please Rinse, Fun...",https://us.shein.com/1pc-Creative-Toilet-Stick...,0,unknown,$0.33,700+ sold recently,76,unknown,0.0,unknown,unknown,unknown
2,"1pc Wall Mounted Hat Rack, Self Adhesive Hat H...",https://us.shein.com/1pc-Wall-Mounted-Hat-Rack...,0,unknown,$1.70,3.7k+ sold recently,15,unknown,0.0,unknown,unknown,unknown
3,"1pc Solar Garden Watering Can Lights,Solar Wat...",https://us.shein.com/1pc-Solar-Garden-Watering...,4,in Path Lights,$4.38,unknown,64,unknown,0.0,unknown,unknown,unknown
4,1000/500/100/50/20pcs Disposable Nitrile Glove...,https://us.shein.com/1000-500-100-50-20pcs-Dis...,0,unknown,$3.10,4.2k+ sold recently,0,unknown,0.0,unknown,unknown,unknown
5,3pcs Creative Powerful Mop Hook Shelf Traceles...,https://us.shein.com/3pcs-Creative-Powerful-Mo...,0,unknown,$0.90,unknown,50,unknown,0.0,unknown,unknown,unknown
6,"16ft-100ft LED Strip Lights,Sync With Music,Wi...",https://us.shein.com/16ft-100ft-LED-Strip-Ligh...,0,unknown,$3.70,10k+ sold recently,0,unknown,0.0,unknown,unknown,unknown
7,"3ft-100ft LED Strip Lights,With IR 44Key Remo...",https://us.shein.com/3ft-100ft-LED-Strip-Light...,9,in Lighting & Lamp,$3.50,unknown,0,unknown,0.0,unknown,unknown,unknown
8,"3pcs Random Color Round Data Cable Organizer, ...",https://us.shein.com/3pcs-Random-Color-Round-D...,0,unknown,$1.28,1.8k+ sold recently,20,unknown,0.0,unknown,unknown,unknown
9,1pc Transparent Stainless Steel Household Scis...,https://us.shein.com/1pc-Transparent-Stainless...,9,in Hand Scissors,$0.51,unknown,76,unknown,0.0,unknown,unknown,unknown


## 📌 Code Summary: Targeted Missing Value Handling

### 1. Numeric Safe Guards
* Targets `color-count` first to fill its missing values (`NaN`) with `0`. This preserves its numeric data type so it doesn't get ruined by string conversions later.

### 2. Explicit Column Targeting
* Instead of doing a blanket fill on the whole dataset, you created a specific list (`text_cols`) containing only your known text and image link columns.

### 3. Isolated Text Cleaning
* Loops through your text list, checks if the column exists in the dataset, fills only its empty spots with `'unknown'`, and converts it to a clean string (`.astype(str)`).

### 4. Why this approach is better:
* It is much safer than a global blanket fill because it protects other columns (like `price` or `selling_proposition`) from accidentally being overwritten with the word `'unknown'` before they are properly cleaned!

In [30]:
# 1. Handle the numeric column first so it stays a proper number
if 'color-count' in df_master.columns:
    df_master['color-count'] = df_master['color-count'].fillna(0)

# 2. The exact column names matching your table perfectly
text_cols = [
    'goods-title-link',
    'goods-title-link--jump',
    'goods-title-link--jump href',
    'rank-sub',
    'blackfridaybelts-bg src',
    'blackfridaybelts-content',
    'product-locatelabels-img src'
]

# 3. Fill missing gaps with 'unknown' while they are still recognized as NaN
for col in text_cols:
    if col in df_master.columns:
        df_master[col] = df_master[col].fillna('unknown').astype(str)

# 4. Print to verify the clean output
df_master.head(10)

,goods-title-link--jump,goods-title-link--jump href,rank-title,rank-sub,price,selling_proposition,discount,goods-title-link,color-count,blackfridaybelts-bg src,blackfridaybelts-content,product-locatelabels-img src
0,46pcs Car Repair Tool Kit: Ratchet Torque Wren...,https://us.shein.com/46pcs-Car-Repair-Tool-Kit...,10,in Hand Tools,$1.10,unknown,0,unknown,0.0,unknown,unknown,unknown
1,"1pc Creative Toilet Sticker, Please Rinse, Fun...",https://us.shein.com/1pc-Creative-Toilet-Stick...,0,unknown,$0.33,700+ sold recently,76,unknown,0.0,unknown,unknown,unknown
2,"1pc Wall Mounted Hat Rack, Self Adhesive Hat H...",https://us.shein.com/1pc-Wall-Mounted-Hat-Rack...,0,unknown,$1.70,3.7k+ sold recently,15,unknown,0.0,unknown,unknown,unknown
3,"1pc Solar Garden Watering Can Lights,Solar Wat...",https://us.shein.com/1pc-Solar-Garden-Watering...,4,in Path Lights,$4.38,unknown,64,unknown,0.0,unknown,unknown,unknown
4,1000/500/100/50/20pcs Disposable Nitrile Glove...,https://us.shein.com/1000-500-100-50-20pcs-Dis...,0,unknown,$3.10,4.2k+ sold recently,0,unknown,0.0,unknown,unknown,unknown
5,3pcs Creative Powerful Mop Hook Shelf Traceles...,https://us.shein.com/3pcs-Creative-Powerful-Mo...,0,unknown,$0.90,unknown,50,unknown,0.0,unknown,unknown,unknown
6,"16ft-100ft LED Strip Lights,Sync With Music,Wi...",https://us.shein.com/16ft-100ft-LED-Strip-Ligh...,0,unknown,$3.70,10k+ sold recently,0,unknown,0.0,unknown,unknown,unknown
7,"3ft-100ft LED Strip Lights,With IR 44Key Remo...",https://us.shein.com/3ft-100ft-LED-Strip-Light...,9,in Lighting & Lamp,$3.50,unknown,0,unknown,0.0,unknown,unknown,unknown
8,"3pcs Random Color Round Data Cable Organizer, ...",https://us.shein.com/3pcs-Random-Color-Round-D...,0,unknown,$1.28,1.8k+ sold recently,20,unknown,0.0,unknown,unknown,unknown
9,1pc Transparent Stainless Steel Household Scis...,https://us.shein.com/1pc-Transparent-Stainless...,9,in Hand Scissors,$0.51,unknown,76,unknown,0.0,unknown,unknown,unknown


In [31]:
df_master

,goods-title-link--jump,goods-title-link--jump href,rank-title,rank-sub,price,selling_proposition,discount,goods-title-link,color-count,blackfridaybelts-bg src,blackfridaybelts-content,product-locatelabels-img src
0,46pcs Car Repair Tool Kit: Ratchet Torque Wren...,https://us.shein.com/46pcs-Car-Repair-Tool-Kit...,10,in Hand Tools,$1.10,unknown,0,unknown,0.0,unknown,unknown,unknown
1,"1pc Creative Toilet Sticker, Please Rinse, Fun...",https://us.shein.com/1pc-Creative-Toilet-Stick...,0,unknown,$0.33,700+ sold recently,76,unknown,0.0,unknown,unknown,unknown
2,"1pc Wall Mounted Hat Rack, Self Adhesive Hat H...",https://us.shein.com/1pc-Wall-Mounted-Hat-Rack...,0,unknown,$1.70,3.7k+ sold recently,15,unknown,0.0,unknown,unknown,unknown
3,"1pc Solar Garden Watering Can Lights,Solar Wat...",https://us.shein.com/1pc-Solar-Garden-Watering...,4,in Path Lights,$4.38,unknown,64,unknown,0.0,unknown,unknown,unknown
4,1000/500/100/50/20pcs Disposable Nitrile Glove...,https://us.shein.com/1000-500-100-50-20pcs-Dis...,0,unknown,$3.10,4.2k+ sold recently,0,unknown,0.0,unknown,unknown,unknown
...,...,...,...,...,...,...,...,...,...,...,...,...
82100,unknown,unknown,9,in Graphic Men Swimwear,$11.39,unknown,7,Manfinity Men's Vacation Zebra-Print Drawstrin...,0.0,unknown,unknown,unknown
82101,unknown,unknown,0,unknown,$13.09,10+ sold recently,14,GAWX Men's Casual Color-Block Slogan Printed T...,0.0,unknown,unknown,unknown
82102,unknown,unknown,8,in Drawstring Men Shirts,$14.29,unknown,23,Manfinity Hypemode Men Pocket Front Drawstring...,0.0,unknown,unknown,unknown
82103,unknown,unknown,6,in Long Sleeve Men Denim Shirts,$25.39,unknown,7,DAZY Men Patched Pocket Denim Shirt,0.0,unknown,unknown,unknown


## 📌 Code Summary: Custom Scale Function (`clean_sales`)

### 1. Pre-Cleaning & Text Standardization
* Forces incoming values to strings, lowers the case, and trims whitespace. If the text matches standard missing values (`'unknown'`, `'nan'`, etc.), it returns `0` immediately.

### 2. Smart Number Extraction
* Uses the regex pattern `([\d\.]+k?)` to grab only the digits, decimal points, and the optional letter `'k'` (e.g., extracting `"3.7k"` out of `"3.7k+ sold recently"`).

### 3. Multiplier Processing (`k` handling)
* Checks if `'k'` is present in the extracted string. If found, it strips the letter, converts the remaining text to a float, multiplies it by `1000`, and casts it to a clean integer (turning `"3.7k"` into `3700`). Otherwise, it casts standard raw digit strings directly into an integer.

### 4. Row-by-Row Mapping (`.apply()`)
* Uses `.apply(clean_sales)` to run this custom logic sequentially across every single row in the `selling_proposition` column, successfully sanitizing your sales volumes!

In [32]:
def clean_sales(val):
    # Ensure it's a string and lowercase for easy matching
    val = str(val).lower().strip()

    # If it's a missing value or unknown, default it to 0
    if val in ['unknown', 'nan', 'none', '']:
        return 0

    # Extract just the numbers and periods/letters using a quick regex split
    # This catches things like '3.7k' or '700'
    match = pd.Series(val).str.extract(r'([\d\.]+k?)')[0].iloc[0]

    if pd.isna(match):
        return 0

    # Handle the 'k' multiplier
    if 'k' in match:
        num = match.replace('k', '')
        return int(float(num) * 1000)
    else:
        return int(float(match))

# Apply the cleaning function to the column
df_master['selling_proposition'] = df_master['selling_proposition'].apply(clean_sales)

# Print to see your perfectly calculated sales numbers!
df_master.head(10)

,goods-title-link--jump,goods-title-link--jump href,rank-title,rank-sub,price,selling_proposition,discount,goods-title-link,color-count,blackfridaybelts-bg src,blackfridaybelts-content,product-locatelabels-img src
0,46pcs Car Repair Tool Kit: Ratchet Torque Wren...,https://us.shein.com/46pcs-Car-Repair-Tool-Kit...,10,in Hand Tools,$1.10,0,0,unknown,0.0,unknown,unknown,unknown
1,"1pc Creative Toilet Sticker, Please Rinse, Fun...",https://us.shein.com/1pc-Creative-Toilet-Stick...,0,unknown,$0.33,700,76,unknown,0.0,unknown,unknown,unknown
2,"1pc Wall Mounted Hat Rack, Self Adhesive Hat H...",https://us.shein.com/1pc-Wall-Mounted-Hat-Rack...,0,unknown,$1.70,3700,15,unknown,0.0,unknown,unknown,unknown
3,"1pc Solar Garden Watering Can Lights,Solar Wat...",https://us.shein.com/1pc-Solar-Garden-Watering...,4,in Path Lights,$4.38,0,64,unknown,0.0,unknown,unknown,unknown
4,1000/500/100/50/20pcs Disposable Nitrile Glove...,https://us.shein.com/1000-500-100-50-20pcs-Dis...,0,unknown,$3.10,4200,0,unknown,0.0,unknown,unknown,unknown
5,3pcs Creative Powerful Mop Hook Shelf Traceles...,https://us.shein.com/3pcs-Creative-Powerful-Mo...,0,unknown,$0.90,0,50,unknown,0.0,unknown,unknown,unknown
6,"16ft-100ft LED Strip Lights,Sync With Music,Wi...",https://us.shein.com/16ft-100ft-LED-Strip-Ligh...,0,unknown,$3.70,10000,0,unknown,0.0,unknown,unknown,unknown
7,"3ft-100ft LED Strip Lights,With IR 44Key Remo...",https://us.shein.com/3ft-100ft-LED-Strip-Light...,9,in Lighting & Lamp,$3.50,0,0,unknown,0.0,unknown,unknown,unknown
8,"3pcs Random Color Round Data Cable Organizer, ...",https://us.shein.com/3pcs-Random-Color-Round-D...,0,unknown,$1.28,1800,20,unknown,0.0,unknown,unknown,unknown
9,1pc Transparent Stainless Steel Household Scis...,https://us.shein.com/1pc-Transparent-Stainless...,9,in Hand Scissors,$0.51,0,76,unknown,0.0,unknown,unknown,unknown


## 📌 Code Summary: Float Extraction & Safe Price Conversion

### 1. Robust Digit & Decimal Extraction
* Converts the `price` column to a string and applies the regex pattern `([\d\.]+)`.
* Unlike `\d+` (which breaks at periods), the `[\d\.]` character class captures both raw numbers and the literal decimal point, ensuring all cents remain perfectly intact (e.g., transforming `"$1.10"` to `"1.10"`).

### 2. Safe Data Type Conversion (`pd.to_numeric`)
* Converts the extracted string decimal directly into a float data type.
* Uses `errors='coerce'` as a failsafe; if a row contains unparseable text, it forces it into a true `NaN` blank instead of crashing the pipeline.

### 3. Missing Value Defaulting (`.fillna(0.0)`)
* Fills any missing or corrupted price values (`NaN`) with a default float value of `0.0`, ensuring data consistency across the entire column.

In [33]:
df_master['price'] = pd.to_numeric(
    df_master['price'].astype(str).str.extract(r'([\d\.]+)', expand=False),
    errors='coerce'
)
df_master['price'] = df_master['price'].fillna(0.0).astype(float)

In [35]:
df_master

,goods-title-link--jump,goods-title-link--jump href,rank-title,rank-sub,price,selling_proposition,discount,goods-title-link,color-count,blackfridaybelts-bg src,blackfridaybelts-content,product-locatelabels-img src
0,46pcs Car Repair Tool Kit: Ratchet Torque Wren...,https://us.shein.com/46pcs-Car-Repair-Tool-Kit...,10,in Hand Tools,1.10,0,0,unknown,0.0,unknown,unknown,unknown
1,"1pc Creative Toilet Sticker, Please Rinse, Fun...",https://us.shein.com/1pc-Creative-Toilet-Stick...,0,unknown,0.33,700,76,unknown,0.0,unknown,unknown,unknown
2,"1pc Wall Mounted Hat Rack, Self Adhesive Hat H...",https://us.shein.com/1pc-Wall-Mounted-Hat-Rack...,0,unknown,1.70,3700,15,unknown,0.0,unknown,unknown,unknown
3,"1pc Solar Garden Watering Can Lights,Solar Wat...",https://us.shein.com/1pc-Solar-Garden-Watering...,4,in Path Lights,4.38,0,64,unknown,0.0,unknown,unknown,unknown
4,1000/500/100/50/20pcs Disposable Nitrile Glove...,https://us.shein.com/1000-500-100-50-20pcs-Dis...,0,unknown,3.10,4200,0,unknown,0.0,unknown,unknown,unknown
...,...,...,...,...,...,...,...,...,...,...,...,...
82100,unknown,unknown,9,in Graphic Men Swimwear,11.39,0,7,Manfinity Men's Vacation Zebra-Print Drawstrin...,0.0,unknown,unknown,unknown
82101,unknown,unknown,0,unknown,13.09,10,14,GAWX Men's Casual Color-Block Slogan Printed T...,0.0,unknown,unknown,unknown
82102,unknown,unknown,8,in Drawstring Men Shirts,14.29,0,23,Manfinity Hypemode Men Pocket Front Drawstring...,0.0,unknown,unknown,unknown
82103,unknown,unknown,6,in Long Sleeve Men Denim Shirts,25.39,0,7,DAZY Men Patched Pocket Denim Shirt,0.0,unknown,unknown,unknown


## Feature Engineering: Extracting Attributes from Titles

### 1. What it is
* Taking a raw string column (like `goods-title-link--jump` containing text like `"1pc Creative Toilet Sticker"`) and pulling out hidden data attributes into their own dedicated columns (like a `pack_size` column containing just `1`).

### 2. How it works (The Logic)
* You use a Regular Expression pattern like `r'^(\d+)(pc|pcs|pack|set)\b'` which scans the very beginning of the product title.
* It grabs the number (`\d+`) and checks if it's followed immediately by a keyword like `pc` (piece) or `pcs`.

### 3. Why this adds high value
* **Price Per Unit:** Knowing the pack size allows you to calculate the *true cost per piece* (e.g., Is a \$3 pack of 10 items cheaper per piece than a \$1 pack of 2 items?).
* **Filtering:** It lets you easily sort or group your 82,105 rows by single items versus multi-packs for deeper e-commerce business insights.

In [37]:
df_master['pack_size'] = df_master['goods-title-link--jump'].str.extract(r'^(\d+)(?:pc|pcs|pack|set)', expand=False)
df_master['pack_size'] = df_master['pack_size'].fillna(1).astype(int)
df_master['pack_size']

/tmp/ipykernel_1179/3784071381.py:2: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_master['pack_size'] = df_master['pack_size'].fillna(1).astype(int)


,pack_size
0,1
1,1
2,1
3,1
4,1
...,...
82100,1
82101,1
82102,1
82103,1


### 1. Enables Granular Business Analytics
* Right now, your dataset has 82,105 rows mixed together. You can't easily answer questions like: *"What is the average discount for Hand Tools vs. Portable Fans?"*
* By isolating clean categories into a `sub_category` column, you can use `.groupby('sub_category')` to instantly compare market trends across different item classes.

### 2. Fixes Messy, Inconsistent Strings
* The raw data has structural inconsistencies because some rows start with `"in "` while others start with `"in Top rated in "`.
* Stripping these prefixes creates standard, clean text keys (e.g., changing `"in Hand Tools"` and `"in Top rated in Hand Tools"` both into just `"Hand Tools"`), preventing your groupings from breaking.

### 3. Powers Market Competitiveness Metrics
* Once you have clean categories combined with the **Original Price** calculation, you can determine which product niches on Shein rely on the most aggressive discount strategies to drive sales.

In [41]:
# 1. Strip out the common ranking prefixes to isolate the clean category name
df_master['sub_category'] = df_master['rank-sub'].str.replace(r'^(in Top rated in |in )', '', regex=True).str.strip()

# 2. View the results to confirm "in Hand Tools" became "Hand Tools"
df_master[['rank-sub', 'sub_category']].head(10)


,rank-sub,sub_category
0,in Hand Tools,Hand Tools
1,unknown,unknown
2,unknown,unknown
3,in Path Lights,Path Lights
4,unknown,unknown
5,unknown,unknown
6,unknown,unknown
7,in Lighting & Lamp,Lighting & Lamp
8,unknown,unknown
9,in Hand Scissors,Hand Scissors


In [48]:
df_master = df_master.copy()

df_master['original_price'] = np.where(
    df_master['discount'] < 100,
    df_master['price'] / (1 - (df_master['discount'] / 100)),
    df_master['price']
)

# 2. Round the new metric to 2 decimal places for clean currency formatting
df_master['original_price'] = df_master['original_price'].round(2)

# 3. View your new financial data structure
df_master[['price', 'discount', 'original_price']].head(10)

,price,discount,original_price
0,1.10,0,1.10
1,0.33,76,1.38
2,1.70,15,2.00
3,4.38,64,12.17
4,3.10,0,3.10
5,0.90,50,1.80
6,3.70,0,3.70
7,3.50,0,3.50
8,1.28,20,1.60
9,0.51,76,2.12


In [49]:
from sqlalchemy import create_engine

# 1. Create a connection engine to a local SQLite database file
# 'sqlite:///...' creates a local file named shein_data.db in your current directory
engine = create_engine('sqlite:///shein_data.db')

# 2. Upload your fully cleaned DataFrame into a database table named 'products'
# if_exists='replace' ensures that if you re-run this cell, it refreshes the table cleanly
df_master.to_sql('products', con=engine, if_exists='replace', index=False)

print("💾 Database setup complete! 82,105 rows successfully written to 'products' table.")

💾 Database setup complete! 82,105 rows successfully written to 'products' table.


In [57]:
from sqlalchemy import create_engine
import pandas as pd

# 1. Re-connect and overwrite the table with your fully updated df_master
engine = create_engine('sqlite:///shein_data.db')
df_master.to_sql('products', con=engine, if_exists='replace', index=False)

# 2. Run the analysis query again now that the columns exist in SQL
query = """
SELECT sub_category, COUNT(*) as total_items, ROUND(AVG(discount), 2) as avg_discount_pct
FROM products
WHERE sub_category != 'unknown'
GROUP BY sub_category
ORDER BY avg_discount_pct DESC
LIMIT 5;
"""

df_sql_result = pd.read_sql(query, con=engine)
df_sql_result

,sub_category,total_items,avg_discount_pct


In [58]:
# Check the first 5 raw rows directly from the SQLite table to see what's inside
df_check = pd.read_sql("SELECT * FROM products LIMIT 5;", con=engine)
df_check

,goods-title-link--jump,goods-title-link--jump href,rank-title,rank-sub,price,selling_proposition,discount,goods-title-link,color-count,blackfridaybelts-bg src,blackfridaybelts-content,product-locatelabels-img src,pack_size,sub_category,original_price
0,46pcs Car Repair Tool Kit: Ratchet Torque Wren...,https://us.shein.com/46pcs-Car-Repair-Tool-Kit...,10,in Hand Tools,1.10,0,0,unknown,0.0,unknown,unknown,unknown,1,Hand Tools,1.10
1,"1pc Creative Toilet Sticker, Please Rinse, Fun...",https://us.shein.com/1pc-Creative-Toilet-Stick...,0,unknown,0.33,700,76,unknown,0.0,unknown,unknown,unknown,1,unknown,1.38
2,"1pc Wall Mounted Hat Rack, Self Adhesive Hat H...",https://us.shein.com/1pc-Wall-Mounted-Hat-Rack...,0,unknown,1.70,3700,15,unknown,0.0,unknown,unknown,unknown,1,unknown,2.00
3,"1pc Solar Garden Watering Can Lights,Solar Wat...",https://us.shein.com/1pc-Solar-Garden-Watering...,4,in Path Lights,4.38,0,64,unknown,0.0,unknown,unknown,unknown,1,Path Lights,12.17
4,1000/500/100/50/20pcs Disposable Nitrile Glove...,https://us.shein.com/1000-500-100-50-20pcs-Dis...,0,unknown,3.10,4200,0,unknown,0.0,unknown,unknown,unknown,1,unknown,3.10


In [59]:
# Run the category aggregation query without the strict count barrier
query = """
SELECT sub_category, COUNT(*) as total_items, ROUND(AVG(discount), 2) as avg_discount_pct
FROM products
WHERE sub_category != 'unknown'
GROUP BY sub_category
ORDER BY avg_discount_pct DESC;
"""

df_sql_result = pd.read_sql(query, con=engine)
df_sql_result

,sub_category,total_items,avg_discount_pct
